From 3ea92b41a8e559bd8d2d2b7c4a287ce77b8cadcf Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Mon, 3 May 2021 21:36:18 -0500
Subject: [PATCH 096/196] Fix getblocktemplate for testnet (#170)

This fixes a bug where when the testnet resets the nBits to 0x1c100000, the coinbase outputs would remain unchanged, resulting in invalid blocks.
---
 src/miner.cpp      |  8 ++++++--
 src/miner.h        |  6 ++++++
 src/rpc/mining.cpp | 16 ++++++++--------
 3 files changed, 20 insertions(+), 10 deletions(-)

diff --git a/src/miner.cpp b/src/miner.cpp
index 76c265f83..0aea977bc 100644
--- a/src/miner.cpp
+++ b/src/miner.cpp
@@ -47,6 +47,11 @@ int64_t UpdateTime(CBlockHeader *pblock, const CChainParams &chainParams,
     return nNewTime - nOldTime;
 }
 
+Amount GetBlockRewardFromFees(Amount nFees) {
+    // 50% of fees get burned.
+    return nFees / 2;
+}
+
 uint64_t CTxMemPoolModifiedEntry::GetVirtualSizeWithAncestors() const {
     return GetVirtualTransactionSize(nSizeWithAncestors,
                                      nSigOpCountWithAncestors);
@@ -185,8 +190,7 @@ BlockAssembler::CreateNewBlock(const CScript &scriptPubKeyIn) {
     }
 
     int64_t nTime1 = GetTimeMicros();
-    // 50% of fees get burned.
-    const Amount amountFeeReward = nFees / 2;
+    const Amount amountFeeReward = GetBlockRewardFromFees(nFees);
 
     m_last_block_num_txs = nBlockTx;
     m_last_block_size = nBlockSize;
diff --git a/src/miner.h b/src/miner.h
index bd4f90701..240f7324a 100644
--- a/src/miner.h
+++ b/src/miner.h
@@ -65,6 +65,12 @@ struct CTxMemPoolModifiedEntry {
     int64_t nSigOpCountWithAncestors;
 };
 
+/**
+ * Calculate the additional block reward in the coinbase from fees.
+ * Currenty, 50% of fees are burned.
+ */
+Amount GetBlockRewardFromFees(Amount nFees);
+
 /**
  * Comparator for CTxMemPool::txiter objects.
  * It simply compares the internal memory address of the CTxMemPoolEntry object
diff --git a/src/rpc/mining.cpp b/src/rpc/mining.cpp
index 5aec17e02..cd9973275 100644
--- a/src/rpc/mining.cpp
+++ b/src/rpc/mining.cpp
@@ -908,14 +908,20 @@ static UniValue getblocktemplate(const Config &config,
     // pointer for convenience
     CBlock *pblock = &pblocktemplate->block;
 
-    // Update nTime
+    // Update nTime (can update nBits on testnet)
     UpdateTime(pblock, chainparams, pindexPrev);
     pblock->nNonce = 0;
 
     UniValue aCaps(UniValue::VARR);
     aCaps.push_back("proposal");
 
-    Amount coinbasevalue = Amount::zero();
+    const Consensus::Params &consensusParams = chainparams.GetConsensus();
+    // CreateNewBlock stores the total fees times -1 in entries[0].fees
+    Amount amountFeeReward =
+        GetBlockRewardFromFees(-1 * pblocktemplate->entries[0].fees);
+    // UpdateTime can update nBits, so we recompute the subsidy
+    Amount coinbasevalue =
+        amountFeeReward + GetBlockSubsidy(pblock->nBits, consensusParams);
 
     UniValue transactions(UniValue::VARR);
     transactions.reserve(pblock->vtx.size());
@@ -926,11 +932,6 @@ static UniValue getblocktemplate(const Config &config,
 
         if (tx.IsCoinBase()) {
             index_in_template++;
-
-            for (const auto &o : pblock->vtx[0]->vout) {
-                coinbasevalue += o.nValue;
-            }
-
             continue;
         }
 
@@ -952,7 +953,6 @@ static UniValue getblocktemplate(const Config &config,
     UniValue aux(UniValue::VOBJ);
 
     UniValue requiredOutputsList(UniValue::VARR);
-    const Consensus::Params &consensusParams = chainparams.GetConsensus();
     for (auto requiredOutput : GetMinerFundRequiredOutputs(
              consensusParams, pindexPrev, coinbasevalue)) {
         UniValue requiredOutputObj(UniValue::VOBJ);
-- 
2.39.5 (Apple Git-154)

